spring-data-neo4j/src/main/java/org/springframework/data/neo4j/context/support/CypherMessageSource.java

```diff
@@ -0,0 +1,162 @@
+import java.util.Map.Entry;
+
+import org.apache.commons.lang.LocaleUtils;
+import org.neo4j.ogm.session.result.Result;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.support.AbstractMessageSource;
+import org.springframework.data.neo4j.template.Neo4jOperations;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+
+/**
+ * <p>
+ * An implementation of {@link org.springframework.context.MessageSource}
+ * which loads messages from a graph database using Cypher, 
+ * supporting basic localization and internationalization.
+ * </p>
+ * <p>
+ * The Cypher query used to retrieve localized messages from the database can be 
+ * overridden by using <code>setQueryCypher()</code>. Messages can be created within 
+ * the database by using Cypher statements such as:
+ * <br/>
+ * <br/>
+ * <blockquote>
+ * <pre>
+ * CREATE (n1:LocalizedMessage { code: 'goodbye', en_US: 'Goodbye', en_GB: 'Cheerio'})
+ * </pre>
+ * </blockquote>
+ * </p>
+ * <p>
+ * Should your application make use of JavaConfig, this class can be registered within 
+ * your <code>ApplicationContext</code> by using the following configuration:
+ * </p>
+ * <blockquote>
+ * <pre>
+ * {@literal @}Bean
+ *   public MessageSource messageSource() {
+ *
+ *        MessageSource cypherMessageSource = new CypherMessageSource();
+ *
+ *        return cypherMessageSource;
+ *
+ *    }
+ * }
+ * </pre>
+ * </blockquote>
+ * @author Eric Spiegelberg - eric [at] miletwentyfour [dot] com
+ */
+@Service
+public class CypherMessageSource extends AbstractMessageSource {
+
+    @Autowired
+    private Neo4jOperations neo4jOperations;
+
+    private boolean initialized;
+
+    private String queryCypher = "match (n:LocalizedMessage) return n";
+
+    private Map<String, Map<Locale, String>> messages = new HashMap<String, Map<Locale, String>>();
+
+    public void initialize() {
+
+        if (!initialized && neo4jOperations != null) {
+
+            initialized = true;
+
+            Map<String, Object> parameters = new HashMap<String, Object>();
+            Result results = neo4jOperations.query(queryCypher, parameters, true);
+
+            for (Map<String, Object> result : results) {
+
+                Iterator<Entry<String, Object>> resultEntry = result.entrySet().iterator();
+
+                while (resultEntry.hasNext()) {
+
+                    Entry<String, Object> node = resultEntry.next();
+
+                    @SuppressWarnings("unchecked")
+                    Map<String, String> nodeProperties = (Map<String, String>) node.getValue();
+
+                    String code = nodeProperties.remove("code");
+
+                    for (Entry<String, String> nodeEntry : nodeProperties.entrySet()) {
+
+                        String text = nodeEntry.getValue();
+
+                        String localeString = nodeEntry.getKey();
+                        Locale locale = LocaleUtils.toLocale(localeString);
```

Comment:
Are you referring to the look up in this particular block of code or the look up of messages through the message source?

While anything that improves performance --- such as caching -- is a good thing, the above block of code is only a one time operation executed when the CypherMessageSource bean is initialized. Given that, is implementing caching here actually beneficial?

That said, if you can provide more context as to what you'd like to see I'll be happy to implement it.